france_connect/clients.py

```diff
@@ -0,0 +1,169 @@
+import logging
+import secrets
+import urllib.parse
+from typing import List, Tuple
+
+import jwt
+import requests
+from jwt import ExpiredSignatureError, InvalidTokenError, PyJWKClient
+
+from .enums import ACRValues, JWTErrorCodes, Scopes
+
+logger = logging.getLogger()
+
+
+class FranceConnect:
+    """
+    FranceConnect Client.
+
+    This class represents a client for interacting with the FranceConnect API.
+    It provides methods for authentication, token exchange, and user information retrieval.
+
+
+    Parameters
+    ----------
+    client_id (str): The client ID for the FranceConnect application.
+    client_secret (str): The client secret for the FranceConnect application.
+    scopes (List[Scopes]): The list of scopes to request during authentication.
+    post_login_callback_url (str): The URL to redirect to after successful login.
+    post_logout_callback_url (str): The URL to redirect to after logout.
+    fc_base_url (str): The base URL of the FranceConnect API.
+    fc_authorization_url (str, optional): The authorization URL of the FranceConnect API.
+    fc_token_url (str, optional): The token URL of the FranceConnect API.
+    fc_jwks_url (str, optional): The JWKS URL of the FranceConnect API.
+    fc_userinfo_url (str, optional): The userinfo URL of the FranceConnect API.
+    fc_logout_url (str, optional): The logout URL of the FranceConnect API.
+    """
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        scopes: List[Scopes],
+        post_login_callback_url: str,
+        post_logout_callback_url: str,
+        fc_base_url: str = "https://fcp-low.integ01.dev-franceconnect.fr",
+        fc_authorization_url: str = "/api/v2/authorize",
+        fc_token_url: str = "/api/v2/token",
+        fc_jwks_url: str = "/api/v2/jwks",
+        fc_userinfo_url: str = "/api/v2/userinfo",
+        fc_logout_url: str = "/api/v2/session/end",
+    ):
+        self.client_id = client_id
+        self.client_secret = client_secret
+        self.scopes = scopes
+        self.post_login_callback_url = post_login_callback_url
+        self.post_logout_callback_url = post_logout_callback_url
+        self.fc_base_url = fc_base_url
+        self.fc_authorization_url = fc_authorization_url
+        self.fc_token_url = fc_token_url
+        self.fc_jwks_url = fc_jwks_url
+        self.fc_userinfo_url = fc_userinfo_url
+        self.fc_logout_url = fc_logout_url
+
+    def generate_nonce(self) -> str:
+        """Generate a random nonce in hexadecimal string (64 characters).."""
+        return secrets.token_hex(48)
+
+    def generate_state(self) -> str:
+        """Generate a random state in hexadecimal string (64 characters)."""
+        return secrets.token_hex(48)
+
+    def get_authentication_url(
+        self, acr_values: List[ACRValues] = [ACRValues.EIDAS1.value], nonce: str = None, state: str = None
+    ) -> Tuple[str, str, str]:
```

Comment:
Do not use mutable objects as a default argument, use `None` instead and set the default value in the method's body.